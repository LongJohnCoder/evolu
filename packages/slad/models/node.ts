import { Brand } from 'utility-types';
import nanoid from 'nanoid';

/**
 * EditorNodeID is string created with nanoid.
 * Editor nodes need UUID for CRDT, React keys, and the other identity operations.
 */
export type EditorNodeID = Brand<string, 'EditorNodeID'>;

export interface EditorNode {
  readonly id: EditorNodeID;
}

export function isEditorNode(value: unknown): value is EditorNode {
  return value != null && typeof (value as EditorNode).id === 'string';
}

export function id(): EditorNodeID {
  return nanoid() as EditorNodeID;
}
