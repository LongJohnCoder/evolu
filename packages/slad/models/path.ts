export type Path = readonly number[];

export const pathsAreEqual = (path1: Path, path2: Path): boolean => {
  if (path1 === path2) return true;
  const { length } = path1;
  if (length !== path2.length) return false;
  for (let i = 0; i < length; i++) if (path1[i] !== path2[i]) return false;
  return true;
};
