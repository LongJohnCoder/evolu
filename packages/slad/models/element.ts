import { unsafeUpdateAt } from 'fp-ts/lib/Array';
import { Children, ReactDOM, ReactNode } from 'react';
import invariant from 'tiny-invariant';
import { $Values } from 'utility-types';
import { SetNodeEditorPathRef } from '../hooks/useSetNodeEditorPathRef';
import { EditorNode, id, isEditorNode } from './node';
import {
  EditorPath,
  editorPathIsEmpty,
  invariantPathIsNotEmpty,
  invariantParentPathAndLastIndex,
  invariantParentPath,
} from './path';
import {
  EditorSelection,
  editorSelectionAsRange,
  editorSelectionIsCollapsed,
} from './selection';
import {
  EditorText,
  editorTextIsBR,
  EditorTextWithOffset,
  invariantIsEditorText,
  isEditorText,
  isEditorTextWithOffset,
} from './text';

/**
 * EditorElement is the base model for all other editor elements.
 */
export interface EditorElement extends EditorNode {
  readonly children: readonly (EditorElementChild)[];
}

export type EditorElementChild = EditorElement | EditorText;

export type RenderEditorElement = (
  element: EditorElement,
  children: ReactNode,
  ref: SetNodeEditorPathRef,
) => ReactNode;

export type MapEditorElement = <T extends EditorElement>(element: T) => T;

export function isEditorElement(value: unknown): value is EditorElement {
  return (
    isEditorNode(value) && Array.isArray((value as EditorElement).children)
  );
}

/**
 * If point.to isEditorTextWithOffset, return editorText. This ensures
 * we can always get EditorElementChild from EditorElementPoint.
 */
export function editorElementPointAsChild(
  point: EditorElementPoint,
): EditorElementChild {
  return isEditorTextWithOffset(point.to) ? point.to.editorText : point.to;
}

export function invariantIsEditorElement(
  value: unknown,
): value is EditorElement {
  invariant(isEditorElement(value), 'Value is not EditorElement.');
  return true;
}

// export type EditorFragment = readonly EditorElementChild[];

interface EditorReactElementFactory<T, P> extends EditorElement {
  readonly tag: T;
  readonly props?: P;
  readonly children: readonly (EditorReactElement | EditorText)[];
}

export type EditorReactElement = $Values<
  {
    [T in keyof ReactDOM]: EditorReactElementFactory<
      T,
      ReturnType<ReactDOM[T]>['props']
    >;
  }
>;

/**
 * EditorPath can be resolved to EditorElement, EditorText, or EditorTextWithOffset.
 */
export type EditorElementPointTo =
  | EditorElement
  | EditorText
  | EditorTextWithOffset;

export function isEditorElementPointTo(
  value: unknown,
): value is EditorElementPointTo {
  if (value == null) return false;
  return (
    isEditorElement(value) ||
    isEditorText(value) ||
    isEditorTextWithOffset(value)
  );
}

/**
 * EditorElementPoint is materialized EditorPath in EditorElement.
 */
export interface EditorElementPoint {
  to: EditorElementPointTo;
  parents: EditorElement[];
}

export function isEditorElementPoint(
  value: unknown,
): value is EditorElementPoint {
  if (value == null) return false;
  const point = value as EditorElementPoint;
  if (!Array.isArray(point.parents)) return false;
  return isEditorElementPointTo(point.to);
}

export function invariantIsEditorElementPoint(
  value: unknown,
): value is EditorElementPoint {
  invariant(isEditorElementPoint(value), 'Value is not EditorElementPoint.');
  return true;
}

/**
 * Map `<div>a</div>` to `{ id: id(), tag: 'div', children: [{ id: id(), text: 'a' }] }` etc.
 */
export function jsxToEditorReactElement(
  element: JSX.Element,
): EditorReactElement {
  const {
    type: tag,
    props: { children = [], ...props },
  } = element;
  const editorChildren = Children.toArray(children).map(child => {
    if (typeof child === 'string') {
      const text: EditorText = { id: id(), text: child };
      return text;
    }
    if (child.type === 'br') {
      const text: EditorText = { id: id(), text: '' };
      return text;
    }
    return jsxToEditorReactElement(child);
  });
  const editorProps = Object.keys(props).length > 0 ? props : undefined;
  return {
    id: id(),
    tag,
    props: editorProps,
    children: editorChildren,
  };
}

/**
 * Like https://developer.mozilla.org/en-US/docs/Web/API/Node/normalize,
 * except strings can be empty. Empty string is considered to be BR.
 */
export function normalizeEditorElement<T extends EditorElement>(element: T): T {
  return {
    ...element,
    ...(element.children
      ? {
          children: element.children.reduce<(EditorElementChild)[]>(
            (array, child) => {
              if (isEditorElement(child))
                return [...array, normalizeEditorElement(child)];
              if (editorTextIsBR(child)) return [...array, child];
              // Always check existence in an array manually.
              // https://stackoverflow.com/a/49450994/233902
              const previousChild = array.length > 0 && array[array.length - 1];
              if (
                previousChild &&
                isEditorText(previousChild) &&
                !editorTextIsBR(previousChild)
              ) {
                array[array.length - 1] = {
                  ...previousChild,
                  text: previousChild.text + child.text,
                };
                return array;
              }
              return [...array, child];
            },
            [],
          ),
        }
      : null),
  };
}

/**
 * Like https://developer.mozilla.org/en-US/docs/Web/API/Node/normalize,
 * except strings can be empty. Empty string is considered to be BR.
 */
export function editorElementIsNormalized({
  children,
}: EditorElement): boolean {
  return !children.some((child, i) => {
    if (!isEditorText(child)) return !editorElementIsNormalized(child);
    if (editorTextIsBR(child)) return false;
    const previous = children[i - 1];
    if (previous && isEditorText(previous) && !editorTextIsBR(previous))
      return true;
    return false;
  });
}

/**
 * Resolve EditorPath on EditorElement to EditorElementPoint or null.
 */
export function editorElementPoint(path: EditorPath) {
  return (element: EditorElement): EditorElementPoint | null => {
    const parents: EditorElementPoint['parents'] = [];
    let to: EditorElementPoint['to'] = element;
    for (let i = 0; i < path.length; i++) {
      const pathIndex = path[i];
      if (isEditorElement(to)) {
        parents.push(to);
        to = to.children[pathIndex];
        if (to == null) return null;
      } else if (isEditorText(to)) {
        const pathContinues = i < path.length - 1;
        if (pathContinues || pathIndex > to.text.length) return null;
        return { parents, to: { editorText: to, offset: pathIndex } };
      }
    }
    return { parents, to };
  };
}

// TODO: Fix types.
// @ts-ignore
export const recursiveRemoveID = element => {
  if (element == null) return element;
  // eslint-disable-next-line no-shadow, @typescript-eslint/no-unused-vars
  const { id, ...objectWithoutID } = element;
  return {
    ...objectWithoutID,
    // @ts-ignore
    children: element.children.map(child => {
      if (isEditorText(child)) {
        // eslint-disable-next-line no-shadow, @typescript-eslint/no-unused-vars
        const { id, ...childWithoutID } = child;
        return childWithoutID;
      }
      return recursiveRemoveID(child);
    }),
  };
};

export function editorElementLens(path: EditorPath) {
  invariantPathIsNotEmpty(path);
  function get(): (element: EditorElement) => EditorElementPoint | null {
    return element => {
      return editorElementPoint(path)(element);
    };
  }

  function set(child: EditorElementChild): MapEditorElement {
    return element => {
      const point = get()(element);
      if (!invariantIsEditorElementPoint(point)) return element;
      const [parentPath, lastIndex] = invariantParentPathAndLastIndex(
        // When the path points to EditorTextWithOffset, get parent path.
        isEditorTextWithOffset(point.to) ? invariantParentPath(path) : path,
      );
      function getUpdatedChildren() {
        if (editorPathIsEmpty(parentPath)) return child;
        const parentPathChild = element.children[lastIndex];
        if (!invariantIsEditorElement(parentPathChild)) return;
        return editorElementLens(parentPath).set(child)(parentPathChild);
      }
      const updatedChildren = getUpdatedChildren();
      return {
        ...element,
        children: unsafeUpdateAt(
          lastIndex,
          updatedChildren,
          element.children as EditorElement[],
        ),
      };
    };
  }

  function modify(
    modifier: (child: EditorElementChild) => EditorElementChild,
  ): MapEditorElement {
    return element => {
      const point = get()(element);
      if (!isEditorElementPoint(point))
        throw new Error(
          'Not defined point in editorElementLens modify. ' +
            'Check whether EditorState selections matches EditorState element.',
        );
      const child = editorElementPointAsChild(point);
      const nextChild = modifier(child);
      return set(nextChild)(element);
    };
  }
  return { get, set, modify };
}

// TODO:
// export function editorElementLenses(selection)

export function deleteContentElement(
  selection: EditorSelection,
): MapEditorElement {
  return element => {
    const range = editorSelectionAsRange(selection);
    const anchorPoint = editorElementPoint(range.anchor)(element);
    const focusPoint = editorElementPoint(range.focus)(element);
    if (!invariantIsEditorElementPoint(anchorPoint)) return element;
    if (!invariantIsEditorElementPoint(focusPoint)) return element;
    // TODO: Handle other cases, with lenses.
    if (
      // Just deleting text on the same EditorTexts.
      isEditorTextWithOffset(anchorPoint.to) &&
      isEditorTextWithOffset(focusPoint.to) &&
      anchorPoint.to.editorText === focusPoint.to.editorText
    ) {
      const parentPath = invariantParentPath(range.anchor);
      const startOffset = anchorPoint.to.offset;
      const endOffset = focusPoint.to.offset;
      return editorElementLens(parentPath).modify(child => {
        if (!invariantIsEditorText(child)) return child;
        return {
          ...child,
          text: child.text.slice(0, startOffset) + child.text.slice(endOffset),
        };
      })(element);
    }
    return element;
  };
}

export function setTextElement(
  text: string,
  selection: EditorSelection,
): MapEditorElement {
  return element => {
    if (editorSelectionIsCollapsed(selection)) {
      return editorElementLens(selection.anchor).modify(child => {
        if (!invariantIsEditorText(child)) return child;
        return { ...child, text };
      })(element);
    }
    return element;
  };
}
