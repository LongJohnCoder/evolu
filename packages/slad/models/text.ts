import invariant from 'tiny-invariant';
import { Predicate, Refinement } from 'fp-ts/lib/function';
import { EditorNode } from './node';

/**
 * EditorText is object because even the two identical texts need own identity.
 * Child index position is not enough for React key and CRDT.
 * Empty string is rendered as BR.
 */
export interface EditorText extends EditorNode {
  readonly text: string;
}

export const isEditorText: Refinement<EditorNode, EditorText> = (
  u,
): u is EditorText => typeof (u as EditorText).text === 'string';

export function invariantIsEditorText(value: EditorNode): value is EditorText {
  invariant(isEditorText(value), 'The value is not EditorText.');
  return true;
}

export const textIsBR: Predicate<string> = text => text.length === 0;

export function editorTextIsBR(editorText: EditorText) {
  return textIsBR(editorText.text);
}
