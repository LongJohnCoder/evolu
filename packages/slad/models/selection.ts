import { editorPathsAreEqual, EditorPath } from './path';

export interface EditorSelection {
  readonly anchor: EditorPath;
  readonly focus: EditorPath;
}

export function editorSelectionIsCollapsed(
  selection: EditorSelection,
): boolean {
  return editorPathsAreEqual(selection.anchor, selection.focus);
}

export function editorSelectionsAreEqual(
  selection1: EditorSelection,
  selection2: EditorSelection,
): boolean {
  if (selection1 === selection2) return true;
  return (
    editorPathsAreEqual(selection1.anchor, selection2.anchor) &&
    editorPathsAreEqual(selection1.focus, selection2.focus)
  );
}
