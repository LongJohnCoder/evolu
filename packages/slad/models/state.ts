import { Optional } from 'utility-types';
import { createElement } from 'react';
import {
  EditorElement,
  EditorDOMElement,
  jsxToEditorDOMElement,
} from './element';
import { EditorSelection, editorSelectionsAreEqual } from './selection';

export interface EditorState<T extends EditorElement = EditorDOMElement> {
  readonly element: T;
  readonly selection: EditorSelection | null;
  readonly hasFocus: boolean;
}

/**
 * Create editor state. By default, the root element is EditorDOMElement.
 */
export function createEditorState<
  T extends EditorState<EditorElement> = EditorState
>({
  element,
  selection = null,
  hasFocus = false,
}: Optional<T, 'hasFocus' | 'selection'>): T {
  return { element, selection, hasFocus } as T;
}

export function createEditorStateWithText({
  text = '',
  selection = null,
  hasFocus = false,
}: {
  text?: string;
  selection?: EditorSelection | null;
  hasFocus?: boolean;
}) {
  return createEditorState({
    element: jsxToEditorDOMElement(
      createElement('div', { className: 'root' }, text),
    ),
    selection,
    hasFocus,
  });
}

// pokud neni, hmm, proc nemuze bejt jen nic?
// editor el nevi jak kreslit, jedine snad jako... ne, nevi
// ok
// element: jsxToEditorDOMElement(<div className="root">a</div>),

export function editorStatesAreEqual<T extends EditorElement>(
  editorState1: EditorState<T> | null,
  editorState2: EditorState<T> | null,
): boolean {
  if (editorState1 === editorState2) return true;
  if (editorState1 == null || editorState2 == null) return false;
  return (
    editorState1.element === editorState2.element &&
    editorState1.hasFocus === editorState2.hasFocus &&
    editorSelectionsAreEqual(editorState1.selection, editorState2.selection)
  );
}
