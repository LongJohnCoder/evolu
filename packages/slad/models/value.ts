import { Immutable } from 'immer';
import { Selection } from './selection';
import { Element, DivElement } from './element';

/**
 * Immutable value describing Editor state.
 */
export interface Value<T extends Element = DivElement> {
  readonly element: Immutable<T>;
  readonly selection?: Selection | undefined;
}
