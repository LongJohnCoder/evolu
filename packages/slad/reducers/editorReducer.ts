import { assertNever } from 'assert-never';
import { Reducer } from 'react';
import { EditorElement } from '../models/element';
import { EditorSelection, editorSelectionsAreEqual } from '../models/selection';
import {
  deleteContent,
  EditorState,
  isEditorStateSelectionValid,
  setText,
} from '../models/state';

/**
 * Various browser actions for updating EditorState.
 */
export type EditorAction =
  | { type: 'focus' }
  | { type: 'blur' }
  | { type: 'selectionChange'; selection: EditorSelection | null }
  | {
      type: 'setEditorStatePartial';
      change: Partial<EditorState<EditorElement>>;
    }
  // beforeinput actions
  | { type: 'setText'; text: string }
  // TODO: To same, imho nepotrebuju a nechci, pro collapsed ani jinak.
  | { type: 'deleteContent'; selection: EditorSelection };

export type EditorReducer<T extends EditorElement = EditorElement> = Reducer<
  EditorState<T>,
  EditorAction
>;

export const editorReducer: EditorReducer = (state, action) => {
  switch (action.type) {
    case 'focus':
      return { ...state, hasFocus: true };

    case 'blur':
      return { ...state, hasFocus: false };

    case 'selectionChange': {
      if (editorSelectionsAreEqual(action.selection, state.selection))
        return state;
      const nextState = { ...state, selection: action.selection };
      // Editor is listening document selection changes even during typing.
      // That's how we can leverage native selection behavior without hacks.
      // But if the user types quickly, editor selection can be updated
      // before Editor element itself which will be detected as missing point.
      // It seems we can safely ignore such temporary invalid selection.
      if (!isEditorStateSelectionValid(nextState)) return state;
      return nextState;
    }

    case 'setEditorStatePartial':
      return { ...state, ...action.change };

    case 'setText':
      return setText(action.text)(state);

    case 'deleteContent':
      return deleteContent(action.selection)(state);

    default:
      return assertNever(action);
  }
};
