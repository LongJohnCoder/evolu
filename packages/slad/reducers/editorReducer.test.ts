import { editorReducer, EditorAction } from './editorReducer';
import { createEditorStateWithText } from '../models/state';
import { recursiveRemoveID } from '../models/element';

const editorState = createEditorStateWithText({ text: '' });

function testEditorReducer(action: EditorAction) {
  const { element, ...nextState } = editorReducer(editorState, action);
  return {
    ...nextState,
    element: recursiveRemoveID(element),
  };
}

test('onFocus', () => {
  expect(testEditorReducer({ type: 'onFocus' })).toMatchSnapshot();
});

test('onBlur', () => {
  expect(testEditorReducer({ type: 'onBlur' })).toMatchSnapshot();
});
