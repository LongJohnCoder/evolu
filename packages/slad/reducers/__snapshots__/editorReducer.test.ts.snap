// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`onBlur 1`] = `
Object {
  "element": Object {
    "children": Array [
      Object {
        "text": "",
      },
    ],
    "props": Object {
      "className": "root",
    },
    "tag": "div",
  },
  "hasFocus": false,
  "selection": null,
}
`;

exports[`onFocus 1`] = `
Object {
  "element": Object {
    "children": Array [
      Object {
        "text": "",
      },
    ],
    "props": Object {
      "className": "root",
    },
    "tag": "div",
  },
  "hasFocus": true,
  "selection": null,
}
`;
